homeassistant/components/template/lock.py

```diff
@@ -47,6 +49,7 @@
         vol.Required(CONF_LOCK): cv.SCRIPT_SCHEMA,
         vol.Required(CONF_UNLOCK): cv.SCRIPT_SCHEMA,
         vol.Required(CONF_VALUE_TEMPLATE): cv.template,
+        vol.Optional(CONF_CODE_FORMAT_TEMPLATE): cv.template,
```

Comment:
Why is this a template instead of a fixed string? Can you give some example of where it's useful to dynamically change the code format?

Reply:
This seemed to be the most straightforward way to support "require no code at all to lock, but a certain code to unlock". Or a well-known short code to lock, but a long (possibly dynamic) one to unlock.
From a system design perspective, if so desired it is always possible to go to a "secure" state (locked) and you need knowledge to go to an "insecure" (unlocked) state. Or have it the other way around if the meaning of "secure" and "insecure" is swapped in a user's context.

This works neatly with the frontend without introducing new attributes like [Template Alarm's code_arm_required](https://www.home-assistant.io/integrations/alarm_control_panel.template/#code_arm_required).

Reply:
OK I see. That makes the PR work around limitations in the lock entity IMHO, something which is not wanted. Please replace the `code_format_template` with a fixed `code_format` and hit the "Ready for review"-button when done.

Reply:
The long story is that I basically wanted to a) virtually secure certain switches/complex setups and b) have something similar to a switch that properly supports in-between-states. Meaning something complex is turned on (which I don't want to happen by accident and not necessarily by anyone walking through our our hallway), and while this is turning on, which can take quite some time, the UI controls display an intermediate state (like the lock that is "unlocking" and then is "unlocked"). The opposite is not true, so while I whish the turning on to be secured, turning off can happen anytime without securing it.

That drove me to template locks, which currently don't support codes at all, and make this available _and_ "templateable." 

As far as I see it, the main question is: does any existing code rely on the (undocumented, AFAIK) fact that a code format property of a lock never changes?  
I guess the answer is no, because there are places where this can obviously occur, Matter being one example. If no one relies on constant code format, why not make it optionally dynamic for template locks (that target advanced users only) as well? And the format can be based on anything, really, not only locks' state.

Another application I have in mind is a dynamic automatic parental control where one can use something while parents are at home, but need to enter a code if they are not. (All this applies to control surfaces that are accessible to anyone in the house).

Reply:
OK, I see, your use case is essentially about (ab)using lock entities for things which are not actually locks.

I think what you want to do is covered here https://github.com/home-assistant/frontend/discussions/8905, although there's no solution in that discussion. My personal opinion is that we should allow users to set a flag on any entity marking it as requiring confirmation, optionally with a PIN. Frontend and voice assistants should be aware of that flag and show a confirmation dialog / verbally ask for confirmation.

For this PR, I would suggest to remove the templating to get the base functionality - adding code format support to template locks merged. A follow-up PR could support templates where we can continue the discussion.

Reply:
It's not more abuse than [using a light to control home theater volume](https://www.home-assistant.io/integrations/light.template/#theater-volume-control) 😉. I always fancy the template-integration to be something special that exactly enables flexibility like this. It's kind of in its DNA.


I am also on board with adding UI-security measures (UI also meaning interactions by voice assistants) to all interactable entities, also with custom messages as described in the linked frontend issue. However, there are three considerations with this idea and your proposed next steps for this PR (all IMO of course):

1. This property (pin required/not required) needs to be templateable; user (or role) specific settings would be nice as I wouldn't bother myself to enter a PIN to enable A/C, for instance, because I usually sometimes kind of tend to know what I do 😉 
2. The place where the check happens should be customizable. This means that I expect the actual check (after the _requirement_ of a PIN/Code has been determined) to be offloaded to something else to be able to also support changing/rolling codes. Only supporting a static setting here would again severely limit flexibility.
3. Introducing a `code_format` for template locks now would solve about half of my issues, but _if_ later we come to the conclusion that it indeed should be templatable, it would either not fit to the template entities config convention (`_template` suffix) or it would create a second property and deprecate the old one and we'd have to think about procedure when both are set and whatnot...


You think a discussion on the discord would provide more insights into what others think about this whole ordeal? Would it even make a difference?